js/reference.js

```diff
@@ -6595,3 +6618,126 @@
             }
         });
     }
+
+    /**
+     *
+     * @param {ERMrest.Reference} reference reference for the association table
+     * @param {Object} prefillObject computed prefill object after extracting the query param and fetching the data from cookie storage
+     * @param {Object} prefillAnnotation
+     */
+    function Prefill (reference, prefillObject, prefillAnnotation) {
+        verify(prefillObject, "'prefillObject' must be defined");
+
+        var self = this;
+        this._reference = reference;
+        this._prefillObject = prefillObject;
+        this._prefillAnnotation = prefillAnnotation;
+
+        // ignore the fks that are simple and their constituent column is system col
+        var nonSystemColumnFks = this._reference.table.foreignKeys.all().filter(function(fk) {
+            return !(fk.simple && module._systemColumns.indexOf(fk.colset.columns[0]) !== -1);
+        });
+
+        // set of foreignkey columns (they might be overlapping so we're not using array)
+        var fkCols = {};
+        nonSystemColumnFks.forEach(function(fk) {
+            fk.colset.columns.forEach(function(col) {
+                fkCols[col] = true;
+            });
+        });
+
+        // the following are ERMrest.ForeignKeyPseudoColumn
+        this.mainColumn;
+        this.leafColumn;
+
+         // NOTE: I think this can be done differently. Leaving this alone for now for the initial implementation
+        reference.columns.forEach(function(column) {
+            // column should be a foreignkey pseudo column
+            if (!column.isForeignKey) return;
+
+            nonSystemColumnFks.forEach(function(fk) {
+                // column and foreign key `.name` property is a hash value
+                if (column.name === fk.name) {
+                    if (prefillObject.fkColumnNames.indexOf(column.name) !== -1) {
+                        self.mainColumn = column;
+                    } else {
+                        // TODO: what if there are multiple "other" fk columns?
+                        if (self.leafColumn) {
+                            // throw error saying "more than 1 foreign key column available"
```

Comment:
You should handle any corner case that you think can cause issues here. Apart from the case that you mentioned, I'm assuming that the composite keys are also not supported by this, right?

If that's the case you should check for them and make sure we're not engaging the chaise UI in these cases.

I don't think you need to throw any errors. Just treat this the cases that we don't want to support now. So for example if `reference.prefill` returns a `null`, it means something was wrong and we should just continue as normal in chaise without using this new UI.